rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala

```diff
@@ -0,0 +1,239 @@
+          }
+        case _ => substitute(exp, env)
+      },*/
+      term.id
+    )
+
+  def rename(term: Par, j: Int): Par =
+    Par(
+      for (send <- term.sends) yield {
+        rename(send, j)
+      },
+      for (recv <- term.receives) yield {
+        rename(recv, j)
+      },
+      for (eval <- term.evals) yield {
+        rename(eval, j)
+      },
+      for (neu <- term.news) yield {
+        rename(neu, j)
+      },
+      for (exp <- term.exprs) yield {
+        rename(exp, j)
+      },
+      term.id
+    )
+
+  def substitute(term: Quote, env: Env): Quote = Quote(substitute(term.p, env))
+
+  def rename(term: Quote, j: Int): Quote = Quote(rename(term.p, j))
+
+  def substitute(term: Channel, env: Env): Either[Channel, Par] =
+    term match {
+      case q @ Quote(_) => Left(substitute(q, env))
+      case ChanVar(b @ _) =>
+        substitute(b, env) match {
+          case Right(Left(quote)) => Left(quote)
+          case Right(Right(par))  => Right(par)
+          case Left(c @ _)        => Left(ChanVar(c))
+        }
+    }
+
+  def rename(term: Channel, j: Int): Channel =
+    term match {
+      case q @ Quote(_)   => rename(q, j)
+      case ChanVar(varue) => ChanVar(rename(varue, j))
+    }
+
+  def substitute(term: Send, env: Env): Send =
+    substitute(term.chan, env) match {
+      case Left(chan) =>
+        Send(
+          chan,
+          term.data map { par =>
+            substitute(par, env)
+          },
+          term.persistent
+        )
+      case Right(par) => throw new Error(s"Cannot send on expression: $par")
+    }
+
+  def rename(term: Send, j: Int): Send =
+    Send(
+      rename(term.chan, j),
+      term.data map { par =>
+        rename(par, j)
+      },
+      term.persistent
+    )
+
+  def substitute(term: Eval, env: Env): Eval =
+    substitute(term.channel, env) match {
+      case Left(_channel) => Eval(_channel)
+      case Right(par)     => throw new Error(s"Cannot dereference process: $par")
+    }
+
+  def rename(term: Eval, j: Int): Eval =
+    Eval(rename(term.channel, j))
+
+  def substitute(term: New, env: Env): New =
+    New(term.count, substitute(term.p, env))
+
+  def rename(term: New, j: Int): New =
+    New(term.count, rename(term.p, j))
+
+  def substitute(term: Receive, env: Env): Receive =
+    Receive(
+      for { (xs, chan) <- term.binds } yield {
+        substitute(chan, env) match {
+          case Left(_chan) => (xs, _chan)
+          case Right(par)  => throw new Error(s"Cannot receive on expression: $par")
+        }
+      },
+      substitute(term.body, env),
+      term.persistent,
+      term.count
+    )
+
+  def rename(term: Receive, j: Int): Receive =
+    Receive(
+      for ((xs, channel) <- term.binds) yield {
+        (xs, rename(channel, j))
+      },
+      rename(term.body, j),
+      term.persistent,
+      term.count
+    )
+
+  def substitute(term: Var, env: Env): Either[Var, Data] =
```

Comment:
Please call this one something different. It makes sense to substitute EVar, but what you're doing here is a little different.
Something like lookupOrDecrement(Var, Env). It makes it more clear that we're at the ground floor.